# Book Recommendation System - Part 1 (EDA)

## Import libraries

In [28]:
# Import libraries
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from scipy import stats

## Import Dataset

In this project, we will use Goodreads dataset. The dataset was originally scraped from the Goodreads API in September 2017 by Zygmunt Zając and updated by Olivier Simard-Hanley. This dataset consist of 5 files which are:
- ratings.csv: contains user ratings for books they read
- books_enriched.csv: contains metadata for each book (book ID, title, authors, year published, etc)
- to-read.csv: contains books marked "to read" by users
- book_tag.csv: contains tags/shelves/genres assigned by users to books
- tag.csv: contains the tag names corresponding to the tag ids in book_tag.csv

I will only use two files which are user ratings and metadata of the books.

In [29]:
pd.options.display.float_format = '{:.2f}'.format
r = pd.read_csv('Dataset/ratings.csv')
b = pd.read_csv('Dataset/books_enriched.csv')

## Data Understanding

### 1. Rating

In [30]:
r.head()

user_id  book_id  rating
0        1      258       5
1        2     4081       4
2        2      260       5
3        2     9296       5
4        2     2318       3

In **dataset r**, there are three columns, which are:
1. `user_id`: user identification number
2. `book_id`: book identification number
3. `rating`: rating given by `user_id`

In [31]:
r.shape

(5976479, 3)

In [32]:
for col in r.columns:
    print(f"Number of {col} is {r[col].nunique()}")

Number of user_id is 53424
Number of book_id is 10000
Number of rating is 5


There are 5,976,479 ratings given by 53,424 people on 10,000 books.

### 2. Books

Now let's explore our book metadata, we can first chek the shape of the data.

In [33]:
b.shape

(10000, 30)

It has 10000 rows and 30 columns. We want to display values in all columns, therefore we will transpose the table and only show first three rows.

In [34]:
b.head()

Unnamed: 0  index                            authors  average_rating  \
0           0      0                ['Suzanne Collins']            4.34   
1           1      1  ['J.K. Rowling', 'Mary GrandPré']            4.44   
2           2      2                ['Stephenie Meyer']            3.57   
3           3      3                     ['Harper Lee']            4.25   
4           4      4            ['F. Scott Fitzgerald']            3.89   

   best_book_id  book_id  books_count  \
0       2767052        1          272   
1             3        2          491   
2         41865        3          226   
3          2657        4          487   
4          4671        5         1356   

                                         description  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...   
1  Harry Potter's life is miserable. His parents ...   
2  About three things I was absolutely positive.\...   
3  The unforgettable novel of a childhood in a sl...   
4  Alternate Cover Edition ISBN: 0743273567 (ISBN...   

                                              genres  goodreads_book_id  ...  \
0  ['young-adult', 'fiction', 'fantasy', 'science...            2767052  ...   
1  ['fantasy', 'fiction', 'young-adult', 'classics']                  3  ...   
2  ['young-adult', 'fantasy', 'romance', 'fiction...              41865  ...   
3  ['classics', 'fiction', 'historical-fiction', ...               2657  ...   
4  ['classics', 'fiction', 'historical-fiction', ...               4671  ...   

  ratings_3 ratings_4  ratings_5 ratings_count  \
0    560092   1481305    2706317       4780653   
1    455024   1156318    3011543       4602479   
2    793319    875073    1355439       3866839   
3    446835   1001952    1714267       3198671   
4    606158    936012     947718       2683664   

                                     small_image_url  \
0  https://images.gr-assets.com/books/1447303603s...   
1  https://images.gr-assets.com/books/1474154022s...   
2  https://images.gr-assets.com/books/1361039443s...   
3  https://images.gr-assets.com/books/1361975680s...   
4  https://images.gr-assets.com/books/1490528560s...   

                                               title  work_id  \
0            The Hunger Games (The Hunger Games, #1)  2792775   
1  Harry Potter and the Sorcerer's Stone (Harry P...  4640799   
2                            Twilight (Twilight, #1)  3212258   
3                              To Kill a Mockingbird  3275794   
4                                   The Great Gatsby   245494   

  work_ratings_count  work_text_reviews_count  \
0            4942365                   155254   
1            4800065                    75867   
2            3916824                    95009   
3            3340896                    72586   
4            2773745                    51992   

                           authors_2  
0                ['Suzanne Collins']  
1  ['J.K. Rowling', 'Mary GrandPré']  
2                ['Stephenie Meyer']  
3                     ['Harper Lee']  
4            ['F. Scott Fitzgerald']  

[5 rows x 30 columns]

In this dataset, there are 30 columns. However, after investigating them, there are several columns that are repeated. For example too many book_id columns, 2 title columns, 2 authors column, details of the number of reviews per rating for each book, etc. In order to make cleaner metadata, I will do data preprocessing.

**Columns in dataset b**:<br>
- Identification number related (`book_id`, `goodreads_book_id`, `best_book_id`, `work_id`, `isbn`, `isbn13`)
- Title related (`original_title`, `title`)
- Authors related (`authors`, `authors_2`)
- Publication year related (`original_publication_year`, `publishDate`: the publication date)
- Rating related (`average_rating`, `ratings_count`: number of review, `work_ratings_count`, `work_text_reviews_count`, `ratings_1`, `ratings_2`, `ratings_3`, `ratings_4`, `ratings_5`)
- Image Url (`image_url`, `small_image_url`)
- `books_count`: number of edition available
- `language_code`: abbreviated language tags for all books
- `genres`: the genre tags taken from the top shelves users have assigned to a book. Only the main Goodreads genres have been retained
- `pages`: the total page count
- `description`: a free text summarizing the book's content
- Others (`Unnamed: 0`, `index`)

## Statistical Summary

Before we decide which columns we will choose, let's explore the columns first. To make it easier, we can divide the columns to numerical columns and categorical columns and then check the statistics of our data.

In [35]:
#divide the columns to numerical and categorical value
nums = b.select_dtypes(include=[np.number]).columns
cats = b.select_dtypes(exclude=[np.number]).columns

In [36]:
b[nums].describe().T

count             mean             std  \
Unnamed: 0                10000.00          4999.50         2886.90   
index                     10000.00          4526.32         2567.92   
average_rating            10000.00             4.00            0.25   
best_book_id              10000.00       5471213.58      7827329.89   
book_id                   10000.00          5000.50         2886.90   
books_count               10000.00            75.71          170.47   
goodreads_book_id         10000.00       5264696.51      7575461.86   
isbn13                     9415.00 9755044298883.46 442861920665.57   
original_publication_year  9979.00          1981.99          152.58   
pages                      9927.00           359.42          216.95   
ratings_1                 10000.00          1345.04         6635.63   
ratings_2                 10000.00          3110.89         9717.12   
ratings_3                 10000.00         11475.89        28546.45   
ratings_4                 10000.00         19965.70        51447.36   
ratings_5                 10000.00         23789.81        79768.89   
ratings_count             10000.00         54001.24       157369.96   
work_id                   10000.00       8646183.42     11751060.82   
work_ratings_count        10000.00         59687.32       167803.79   
work_text_reviews_count   10000.00          2919.96         6124.38   

                                   min              25%              50%  \
Unnamed: 0                        0.00          2499.75          4999.50   
index                             0.00          2351.75          4567.50   
average_rating                    2.47             3.85             4.02   
best_book_id                      1.00         47911.75        425123.50   
book_id                           1.00          2500.75          5000.50   
books_count                       1.00            23.00            40.00   
goodreads_book_id                 1.00         46275.75        394965.50   
isbn13                    195170342.00 9780316192995.00 9780451528640.00   
original_publication_year     -1750.00          1990.00          2004.00   
pages                             0.00           250.00           336.00   
ratings_1                        11.00           196.00           391.00   
ratings_2                        30.00           656.00          1163.00   
ratings_3                       323.00          3112.00          4894.00   
ratings_4                       750.00          5405.75          8269.50   
ratings_5                       754.00          5334.00          8836.00   
ratings_count                  2716.00         13568.75         21155.50   
work_id                          87.00       1008841.00       2719524.50   
work_ratings_count             5510.00         15438.75         23832.50   
work_text_reviews_count           3.00           694.00          1402.00   

                                       75%              max  
Unnamed: 0                         7499.25          9999.00  
index                              6661.25          9999.00  
average_rating                        4.18             4.82  
best_book_id                    9636112.50      35534230.00  
book_id                            7500.25         10000.00  
books_count                          67.00          3455.00  
goodreads_book_id               9382225.25      33288638.00  
isbn13                    9780830777175.00 9790007672390.00  
original_publication_year          2011.00          2017.00  
pages                               424.00          5216.00  
ratings_1                           885.00        456191.00  
ratings_2                          2353.25        436802.00  
ratings_3                          9287.00        793319.00  
ratings_4                         16023.50       1481305.00  
ratings_5                         17304.50       3011543.00  
ratings_count                     41053.50       4780653.00  
work_id                        14

- There are many unique columns: `Unnamed: 0`, `index`, `best_book_id`, `book_id`, `goodreads_book_id`, `isbn13`, `work_id`. The column `book_id` will be used to merge to rating table, so the other columns will be dropped.

- The oldest book was written in -1750 which means 1750 BC.
- Mean of `average_rating` is 4.002 (pretty good!)
- We found weird number of pages (0), we found that book should have at least 80 pages to be published
- The most popular book has 4,780,653 reviews, the least popular has 2,716 reviews.

In [37]:
b[cats].describe().T

count unique  \
authors          10000   5057   
description       9943   9772   
genres           10000   4351   
image_url        10000   6669   
isbn              9300   9300   
language_code    10000     41   
original_title    9415   9274   
publishDate       9992   4679   
small_image_url  10000   6669   
title            10000   9964   
authors_2        10000   5057   

                                                               top  freq  
authors                                           ['Stephen King']    58  
description      The first in a spectacularly genre-mashing adv...     4  
genres                                      ['fantasy', 'fiction']   152  
image_url        https://s.gr-assets.com/assets/nophoto/book/11...  3332  
isbn                                                     439023483     1  
language_code                                                  eng  9680  
original_title                                                         5  
publishDate                                     (None, None, None)    75  
small_image_url  https://s.gr-assets.com/assets/nophoto/book/50...  3332  
title                                               Selected Poems     4  
authors_2                                         ['Stephen King']    58

- The most popular author is Stephen King
- Selected Poems appears 4 times. Is it the same book?
- Because we already has `original_publication_year`, `publishDate` will be dropped
- `authors`, `genres`, `publishDate`, and `authors_2` has unnecessary characters

## Data Preprocessing

Our next step is data preprocessing. In building machine learning models, we should check for missing values and duplicated rows in our data. In this process we should also transform data into a format that is easier and more effective to process, so that our model will produce more accurate results.

### Missing Value and Duplicated Rows

In [38]:
#check missing value
dataset = [r, b]
for data in dataset:
    print(data.isnull().values.any())

False
True


In [39]:
#check duplicated rows
for data in dataset:
    print(data.duplicated().values.any())

False
False


In [40]:
b.isnull().sum()

Unnamed: 0                     0
index                          0
authors                        0
average_rating                 0
best_book_id                   0
book_id                        0
books_count                    0
description                   57
genres                         0
goodreads_book_id              0
image_url                      0
isbn                         700
isbn13                       585
language_code                  0
original_publication_year     21
original_title               585
pages                         73
publishDate                    8
ratings_1                      0
ratings_2                      0
ratings_3                      0
ratings_4                      0
ratings_5                      0
ratings_count                  0
small_image_url                0
title                          0
work_id                        0
work_ratings_count             0
work_text_reviews_count        0
authors_2                      0
dtype: int

In [41]:
#make a copy 
books = b.copy()

In [42]:
percent_missing = books.isnull().sum() * 100 / len(books)
missing_value = (pd.DataFrame({'percent_missing': percent_missing})).sort_values(by=['percent_missing'], ascending=False)
missing_value[missing_value['percent_missing']>0]

percent_missing
isbn                                  7.00
original_title                        5.85
isbn13                                5.85
pages                                 0.73
description                           0.57
original_publication_year             0.21
publishDate                           0.08

- Impute `original_publication_year` by using `publishDate`, then drop `publishDate`. I chose `original_publication_year` because it has the same format meanwhile `publishDate` has various data styles.
- Impute `pages` with median.
- Impute `description` with book's title
- Drop `isbn`, `original_title`, `isbn13`, there is no need to impute these columns.

In [43]:
#compare original_publication_year and publishdate when original_publication_year is null
books[books['original_publication_year'].isnull()][['original_publication_year', 'publishDate']]

original_publication_year           publishDate
198                         NaN     October 28th 2008
3249                        NaN     February 7th 2006
3881                        NaN   September 29th 2009
4252                        NaN    November 10th 2010
4303                        NaN        June 23rd 2009
4392                        NaN        April 8th 2013
5001                        NaN     November 9th 2004
5208                        NaN     December 6th 2010
5648                        NaN     October 11th 2006
7163                        NaN    November 25th 2004
7850                        NaN     October 15th 2007
8119                        NaN                  2009
8227                        NaN    ('6', '1', '1998')
8532                        NaN   ('11', '5', '2003')
8568                        NaN   ('11', '5', '2003')
9140                        NaN   ('10', '2', '2010')
9151                        NaN    ('2', '2', '2016')
9198                        NaN    ('9', '1', '2001')
9254                        NaN  (None, None, '2000')
9728                        NaN   ('9', '25', '2012')
9842                        NaN   ('6', '16', '2009')

Lucky that all rows that are invalid in `original_publication_year` has the year in `publishDate`. Therefore, we can get year in `publishDate` and impute it to `original_publication_year`.

In [44]:
#get year from publishDate to original_publication_year
books['publishDate'].replace(r"[(')]", "", regex=True, inplace=True)
books['publishDate'] = books['publishDate'].str.extract('(\d{4})$').fillna('')
books['original_publication_year'] = books['original_publication_year'].fillna(books['publishDate'])
books['description'] = books['description'].fillna(books['title'])

In [45]:
#impute null in pages with median
books['pages'] = books['pages'].fillna(books['pages'].median())

In [46]:
#drop unnecessary columns with null
books = books.drop(['isbn', 'original_title', 'isbn13', 'publishDate'], axis=1)

In [47]:
#check missing value
books.isnull().values.any()

False

### Feature Engineering

In [48]:
# Keep important columns, drop the rest

#keep important columns, drop the rest
cols_to_keep = ['book_id', 'title', 'authors', 'original_publication_year', 'pages', 'description', 'genres', 'average_rating', 'ratings_count', 'books_count','small_image_url']
books = books[cols_to_keep]

In [49]:
#delete unnecessary characters from authors, genres and description column
col_trans = ['authors', 'genres', 'description']
for col in col_trans:
    books[col].replace(r"[\[\]\-\\\/\"—'()|:]", "", regex=True, inplace=True) 
books['description'].replace({"\n" : " ", "isbn13" : "", "isbn" : "", r"[0-9]{8,}" : "", r"[.,]" : ""}, regex=True, inplace=True)
books['description'] = books['description'].str.lower()

In [50]:
#rename column and change year, pages, and book_count to integer
books.rename(columns = {'original_publication_year':'year'}, inplace = True)
books[['year', 'pages', 'books_count']] = books[['year', 'pages', 'books_count']].astype(int)

In [51]:
books.sample(10)

book_id                                              title  \
899       959            A Portrait of the Artist as a Young Man   
9062     6922                                     Leading Change   
954      1022                 Heir of Fire (Throne of Glass, #3)   
5200     5861       Appointment with Death (Hercule Poirot, #19)   
5383     6095      Essentialism: The Disciplined Pursuit of Less   
3587     3904       Anne of Ingleside (Anne of Green Gables, #6)   
5277     5962  If You're Reading This, It's Too Late (Secret,...   
2742     2947  Dragons of Winter Night (Dragonlance: Chronicl...   
5991     6877                              A House for Mr Biswas   
5962     6839                                      Suicide Notes   

                               authors  year  pages  \
899          James Joyce, Seamus Deane  1916    329   
9062                    John P. Kotter  1988    208   
954                      Sarah J. Maas  2014    565   
5200                   Agatha Christie  1938    303   
5383                      Greg McKeown  2014    260   
3587                   L.M. Montgomery  1939    274   
5277  Pseudonymous Bosch, Gilbert Ford  2008    385   
2742      Margaret Weis, Tracy Hickman  1985    358   
5991                      V.S. Naipaul  1961    623   
5962               Michael Thomas Ford  2008    295   

                                            description  \
899   the portrayal of stephen dedaluss dublin child...   
9062  john kotter’s nowlegendary eightstep process f...   
954   celaena has survived deadly contests and shatt...   
5200  among the towering red cliffs of petra like so...   
5383  have you ever found yourself stretched too thi...   
3587  anne is the mother of five with never a dull m...   
5277  beware!dangerous secrets lie between the pages...   
2742  they won their first real battle in the war fo...   
5991  mohun biswas has spent his 46 years of life st...   
5962  im not crazy i dont see what the big deal is a...   

                                                 genres  average_rating  \
899                                   classics, fiction            3.60   
9062         business, nonfiction, selfhelp, psychology            4.01   
954   fantasy, youngadult, romance, fiction, paranormal            4.53   
5200        mystery, fiction, crime, classics, thriller            3.84   
5383  nonfiction, selfhelp, business, psychology, ph...            3.99   
3587  classics, fiction, youngadult, historicalficti...            4.05   
5277              mystery, fantasy, fiction, youngadult            4.23   
2742                       fantasy, fiction, youngadult            4.13   
5991                    fiction, classics, contemporary            3.82   
5962      youngadult, contemporary, fiction, psychology            3.92   

      ratings_count  books_count  \
899           92569          766   
9062          12245           25   
954          123843           46   
5200          20793          210   
5383          16137           29   
3587          36641          166   
5277          18310           34   
2742          31341           60   
5991          13073           54   
5962          14749           10   

                                        small_image_url  
899   https://s.gr-assets.com/assets/nophoto/book/50...  
9062  https://s.gr-assets.com/assets/nophoto/book/50...  
954   https://images.gr-assets.com/books/1460846511s...  
5200  https://images.gr-assets.com/books/1308808730s...  
5383  https://images.gr-assets.com/books/1403165375s...  
3587  https://images.gr-assets.com/books/1285712495s...  
5277  https://images.gr-assets.com/books/1344268358s...  
2742  https://s.gr-assets.com/assets/nophoto/book/50...  
5991  https://s.gr-assets.com/assets/nophoto/book/50...  
5962  https://images.gr-assets.com/books/1424990334s...

In [53]:
books.to_csv('books_cleaned.csv')